mujoco_py/mjpid.pyx

```diff
@@ -24,77 +81,191 @@ cdef enum USER_DEFINED_ACTUATOR_PARAMS:
 
+cdef mjtNum c_pi_cascade_bias(const mjModel*m, const mjData*d, int id):
+    """
+    A cascaded PID controller implementation that can control position
+    and velocity setpoints for a given actuator.
+    
+    Currently, the cascaded controller is implemented as PI controllers for both the position and 
+    the velocity terms and therefore require Kp, Ti and max_clamp_integral parameters to be 
+    specified for both. Additionally, an exponential moving average smoothing is applied to the 
+    velocity component for added stability, and the controller is gravity compensated via the 
+    `qfrc_bias` term.
 
-cdef mjtNum c_pid_bias(const mjModel* m, const mjData* d, int id) with gil:
+    These are provided as part of gainprm in the following order: `gainprm="Kp_pos Ti_pos 
+    max_clamp_pos Kp_vel Ti_vel max_clamp_vel ema_smooth_factor max_vel"`, where ema_smooth_factor 
+    denotes the EMA smoothing factor and max_vel is a velocity constraint applied to the velocity setpoint. 
+    """
     cdef double dt_in_sec = m.opt.timestep
-    cdef double error = d.ctrl[id] - d.actuator_length[id]
     cdef int NGAIN = int(const.NGAIN)
 
-    cdef double Kp = m.actuator_gainprm[id * NGAIN + IDX_PROPORTIONAL_GAIN]
-    cdef double error_deadband = m.actuator_gainprm[id * NGAIN + IDX_ERROR_DEADBAND]
-    cdef double integral_max_clamp = m.actuator_gainprm[id * NGAIN + IDX_INTEGRAL_MAX_CLAMP]
-    cdef double integral_time_const = m.actuator_gainprm[id * NGAIN + IDX_INTEGRAL_TIME_CONSTANT]
-    cdef double derivative_gain_smoothing = \
-        m.actuator_gainprm[id * NGAIN  + IDX_DERIVATIVE_GAIN_SMOOTHING]
-    cdef double derivate_time_const = m.actuator_gainprm[id * NGAIN + IDX_DERIVATIVE_TIME_CONSTANT]
+    cdef double Kp_cas = m.actuator_gainprm[id * NGAIN + IDX_CAS_PROPORTIONAL_GAIN]
 
-    cdef double effort_limit_low = m.actuator_forcerange[id * 2]
-    cdef double effort_limit_high = m.actuator_forcerange[id * 2 + 1]
+    if Kp_cas != 0:
+        # Run a position PID loop and use the result to set a desired velocity signal
+        pos_output = _pid(parameters=PIDParameters(
+            dt_seconds=dt_in_sec,
+            setpoint=d.ctrl[id],
+            feedback=d.actuator_length[id],
+            Kp=Kp_cas,
+            error_deadband=0.0,
+            integral_max_clamp=m.actuator_gainprm[id * NGAIN + IDX_CAS_INTEGRAL_MAX_CLAMP],
+            integral_time_const=m.actuator_gainprm[id * NGAIN + IDX_CAS_INTEGRAL_TIME_CONSTANT],
+            derivative_gain_smoothing=0.0,
+            derivative_time_const=0.0,
+            previous_errors=PIDErrors(
+                error=0.0,
+                derivative_error=0.0,
+                integral_error=d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_INTEGRAL_ERROR],
+            )))
 
-    if fabs(error) < error_deadband:
-        error = 0.0
+        # Save errors
+        d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_LAST_ERROR] = pos_output.errors.error
```

Comment:
not used?

Reply:
will rm. also reduces NUM_USER_DATA_PER_ACT to 3 in this case as well 🎉